symeig for variables

ncullen93 opened the pull request:
this implements the symeig function for variables... I grabbed the gradient implementation directly from [theano](https://github.com/Theano/theano/blob/52903f8267cff316fc669e207eac4e2ecae952a6/theano/tensor/nlinalg.py#L432-L461) ... And I test it in [this gist](https://gist.github.com/ncullen93/9acefab137976712de0a51d88b39ffe7) to show that it gets the same gradient as the theano version and that it passes `gradcheck`. I've also used this in actual networks so I'm fairly confident it's correct.

I obviously need to add actual tests, so any direction for that and anything else is much appreciated.

Commit:
symeig for variables

fmassa commented on torch/autograd/_functions/linalg.py:
```diff
@@ -102,3 +102,42 @@ def backward(ctx, grad_output, grad_LU=None):
+        x, w, v, = ctx.saved_variables
+
+        # gives an error if I don't do this..
+        x = x.data
```
This doesn't look right. If you need tensors in your backward, you should use the devorator `once_differentiable` in the function, but looking at the functions used, that shouldn't be necessary.

ncullen93 replied:
yeah this is where I'm confused. I don't understand why I have to call `.data` when no other functions do. This does need `x`, `w`, and `v` for the backward.

fmassa commented on torch/autograd/_functions/linalg.py:
```diff
@@ -102,3 +102,42 @@ def backward(ctx, grad_output, grad_LU=None):
+            tri1 = lambda a: torch.triu(a, 1)
+
+        def G(n):
+            return sum([v[:, m] * grad_v.t()[n].matmul(v[:, m]) / (w[n] - w[m])
```
Can't we simplify that by using sopething like `baddbmm` or some matrix manipulation? This will be very expensive on the GPU as is

ncullen93 replied:
yes definitely, but I figure this will be a drop-in replacement once we sort out the other details above.

fmassa commented on torch/autograd/_functions/linalg.py:
```diff
@@ -102,3 +102,42 @@ def backward(ctx, grad_output, grad_LU=None):
+            return sum([v[:, m] * grad_v.t()[n].matmul(v[:, m]) / (w[n] - w[m])
+                       for m in range(N) if m != n])
+
+        g = sum([torch.ger(v[:, n], v[:, n] * grad_w[n] + G(n))
```
Same here, can't we use a single matrix op to avoid the for loop?

fmassa commented on torch/autograd/_functions/linalg.py:
```diff
@@ -102,3 +102,42 @@ def backward(ctx, grad_output, grad_LU=None):
+
+    @staticmethod
+    def forward(ctx, input, eigenvectors=False, upper=True):
+        ctx.eigenvectors = eigenvectors
```
Do you use it elsewhere? Also, what happens if `eigenvectors=False`? The backward requires then, right?

ncullen93 replied:
no, i dont use it elsewhere atm - it's there for your second point on what to do if `eigenvectors=False`. Seems like you shouldn't be able to calculate the backward if so. Should I raise an exception in `backward` if it's called with `eigenvectors=False` ?

fmassa reviewed (changes requested):
Thanks for the PR! Do you think it would be possible to get `eig` as well?
I made some comments that I think need to be addressed before this PR can be merged. You could get the template for the tests from the gesv or inverse functions (I can give them to you tomorrow if needed, typing from the phone)

ncullen93 commented:
Major questions right now:

1. What to do if `eigenvectors=False`. I think an exception should be raised, because it's not clear how to compute the gradient without it.

2. Why do I have to call `.data` on `x`, `w`, and `v`? Am I saving those for the backend correctly?

Commit:
Update linalg.py

formatting for pyflakes

apaszke commented:
1. You should probably override the flag and save the result for backward.
2. They're saved correctly, but something must be wrong with your `backward` implementation. What error do you get when you don't unpack the variables?

apaszke commented:
Ad 2. You can use `@once_differentiable` and use `ctx.saved_tensors`. It will work, but won't be differentiable twice

Commit:
Update variable.py

small typo

Commit:
Update linalg.py

removed `.data` call (not needed) and renamed `symeig` return values to `e` and `V` instead of `w` and `x` to be consistent with pytorch documentation.

ncullen93 commented:
Re: 1. Ok, great that's a good solution. I will work on that.

Re: 2. Well I just tested it again - in the codebase, it only works without `.data`. However, implemented stand-alone (i.e. implemented as recommended [here](http://pytorch.org/docs/master/notes/extending.html)) it only works <i>with</i> `.data`. Weird. I updated to reflect this.

I get the standard error:
```python
TypeError: torch.dot received an invalid combination of arguments - got (torch.FloatTensor, !Variable!), but expected (torch.FloatTensor source, torch.FloatTensor tensor)
```

apaszke commented:
It seems that the only problem is that you're mixing tensors with Variables, so just making sure that you keep everything wrapped would be enough. Add a couple of prints and see why do they get mixed up

zou3519 commented:
I think this is superceded by https://github.com/pytorch/pytorch/pull/8586; please reopen if I am incorrect.
